### PR TITLE
Reuse empty string & space string logic for event types in devtools

### PR DESCRIPTION
**Pull request by t3chguy**
Fixes https://github.com/vector-im/element-web/issues/23115

image

---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * Reuse empty string & space string logic for event types in devtools ([\#9218](https://github.com/matrix-org/matrix-react-sdk/pull/9218)). Fixes vector-im/element-web#23115.

### CODE REVIEW

File: src/components/views/dialogs/devtools/RoomState.tsx [hunk 1]
```diff
@@ -41,52 +41,61 @@ export const StateEventEditor = ({ mxEvent, onBack }: IEditorProps) => {
...
+    const trimmed = label.trim();
+
+    return <button
+        className={classNames("mx_DevTools_button", {
+            mx_DevTools_RoomStateExplorer_button_hasSpaces: trimmed.length !== label.length,
+            mx_DevTools_RoomStateExplorer_button_emptyString: !trimmed,
+        })}
+        onClick={onClick}
+    >
+        { trimmed ? label : _t("<%(count)s spaces>", { count: label.length }) }
```
[the remark this answers is not in the document]
Comment (t3chguy) [reply src/components/views/dialogs/devtools/RoomState.tsx:59]: Sure, feel free to open a new issue for that, it isn't a new issue

File: src/components/views/dialogs/devtools/RoomState.tsx [hunk 2]
```diff
@@ -41,52 +41,61 @@ export const StateEventEditor = ({ mxEvent, onBack }: IEditorProps) => {
...
+            setEvent(events.get("")!);
         } else {
             setEvent(null);
         }
     }, [events]);
 
     if (event) {
-        const onBack = () => {
-            setEvent(null);
+        const _onBack = () => {
```
[the remark this answers is not in the document]
Comment (t3chguy) [reply src/components/views/dialogs/devtools/RoomState.tsx:83]: You can't use `useCallback` inside an if statement.
https://reactjs.org/docs/hooks-rules.html#:~:text=Don't%20call%20Hooks%20inside,each%20time%20a%20component%20renders.

File: src/components/views/dialogs/devtools/RoomState.tsx [hunk 3]
```diff
@@ -41,52 +41,61 @@ export const StateEventEditor = ({ mxEvent, onBack }: IEditorProps) => {
...
-                            mx_DevTools_RoomStateExplorer_button_emptyString: !trimmed,
-                        })}
-                        key={stateKey}
-                        onClick={onClick}
-                    >
-                        { trimmed ? stateKey : _t("<%(count)s spaces>", { count: stateKey.length }) }
-                    </button>;
-                })
+                Array.from(events.entries()).map(([stateKey, ev]) => (
+                    <StateEventButton key={stateKey} label={stateKey} onClick={() => setEvent(ev)} />
```
[the remark this answers is not in the document]
Comment (t3chguy) [reply src/components/views/dialogs/devtools/RoomState.tsx:97]: You can't use `useCallback` in a loop - https://reactjs.org/docs/hooks-rules.html#:~:text=Don't%20call%20Hooks%20inside,each%20time%20a%20component%20renders.

File: src/components/views/dialogs/devtools/RoomState.tsx [hunk 4]
```diff
@@ -113,19 +122,9 @@ export const RoomStateExplorer = ({ onBack, setTool }: IDevtoolsProps) => {
...
-                    return <button
-                        className="mx_DevTools_button"
-                        key={eventType}
-                        onClick={onClick}
-                    >
-                        { eventType }
-                    </button>;
-                })
+                Array.from(events.keys()).map((eventType) => (
+                    <StateEventButton key={eventType} label={eventType} onClick={() => setEventType(eventType)} />
```
[the remark this answers is not in the document]
Comment (t3chguy) [reply src/components/views/dialogs/devtools/RoomState.tsx:126]: You can't use `useCallback` in a loop - https://reactjs.org/docs/hooks-rules.html#:~:text=Don't%20call%20Hooks%20inside,each%20time%20a%20component%20renders.

File: src/components/views/dialogs/devtools/RoomState.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (t3chguy) [reply src/components/views/dialogs/devtools/RoomState.tsx:59]: This isn't a new issue though

**Review by turt2live — APPROVED** [review]
generally seems fine - as a quality of life thing, this is an improvement.

File: src/components/views/dialogs/devtools/RoomState.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (jaller94) [reply src/components/views/dialogs/devtools/RoomState.tsx:97]: You can define a new functional component outside this functional component. :)

**Pull request merged**